### PR TITLE
Fix deletion crash

### PR DESCRIPTION
**Pull request by zegnus**
**Problem**
1. Queue 10 batches
2. Pause one
3. Delete 7 (including the downloading and the paused one)
4. Crash

**Solution**
The problem was detected when the deletion is not perfectly timed with the next batch to be processed. The next batch will start requesting the fileSize and storing it locally, but at the same time, the batch database is going to be deleted. At the subsequent fileSize write to the database, the parent batch does not exist and the constraint in the database fails (as expected), as you cannot insert a file without an existing parent batch.

The solution is to not delete immediately the batch on delete request, but let the batch process as normal and delete at the end.

There is also the `pause` case. In this scenario the paused batch finishes and no longer exists in the executor. For this scenario, an immediate delete must be done on request.

### CODE REVIEW

File: library/build.gradle [hunk 1]
```diff
@@ -51,7 +51,7 @@ publish {
     userOrg = 'novoda'
     groupId = 'com.novoda'
     artifactId = 'download-manager'
-    publishVersion = 'SNAPSHOT-2.0.27'
+    publishVersion = 'SNAPSHOT-2.0.28'
```
Comment (MechaRyRy) [inline library/build.gradle:54]: 👍

File: library/src/main/java/com/novoda/downloadmanager/DownloadBatch.java [hunk 2]
```diff
@@ -207,7 +222,10 @@ void resume() {
     }
 
     void delete() {
-        downloadsBatchPersistence.deleteAsync(downloadBatchStatus.getDownloadBatchId());
+        if (downloadBatchStatus.status() == PAUSED) {
+            downloadsBatchPersistence.deleteAsync(downloadBatchStatus.getDownloadBatchId());
```
Comment (danybony) [inline library/src/main/java/com/novoda/downloadmanager/DownloadBatch.java:225]: Just to check, do we still need this to be async?

Comment (zegnus) [reply library/src/main/java/com/novoda/downloadmanager/DownloadBatch.java:225]: yes in this case, because this `delete()` is an immediate request (main thread)

File: library/src/main/java/com/novoda/downloadmanager/DownloadsBatchPersistence.java [hunk 3]
```diff
@@ -44,15 +44,14 @@ void persistAsync(DownloadBatchTitle downloadBatchTitle,
...
-                for (DownloadFile downloadFile : downloadFiles) {
-                    downloadFile.persistSync();
-                }
-
                 downloadsPersistence.transactionSuccess();
             } finally {
                 downloadsPersistence.endTransaction();
             }
+
+            for (DownloadFile downloadFile : downloadFiles) {
```
Comment (MechaRyRy) [inline library/src/main/java/com/novoda/downloadmanager/DownloadsBatchPersistence.java:52]: Why was this moved?

Comment (zegnus) [reply library/src/main/java/com/novoda/downloadmanager/DownloadsBatchPersistence.java:52]: now the `downloadFile.persistSync` has his own transaction operations

**Review by danybony — APPROVED** [review]

File: library/src/main/java/com/novoda/downloadmanager/DownloadsBatchPersistence.java (hunk 3 above)
Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/DownloadsBatchPersistence.java:52]: I was wondering if we should move this transaction logic to the `DownloadsPersistence` rather than having it everywhere else.

Comment (MechaRyRy) [reply library/src/main/java/com/novoda/downloadmanager/DownloadsBatchPersistence.java:52]: Downside would be that any new `DownloadPersistence` would have to pay attention to implement the transactions too 🤔

**Review by MechaRyRy — APPROVED** [review]
